File: pulse-react-native-otel/ios/PulseReactNativeOtelTracer.swift
import Foundation
import PulseKit
import OpenTelemetryApi

@objc(PulseReactNativeOtelTracer)
public class PulseReactNativeOtelTracer: NSObject {

    private static func getTracer() -> Tracer {
        if let otel = Pulse.shared.getOtelOrNull() {
            return otel.tracerProvider.get(
                instrumentationName: PulseOtelConstants.INSTRUMENTATION_SCOPE,
                instrumentationVersion: PulseOtelConstants.INSTRUMENTATION_VERSION
            )
        }
        return DefaultTracerProvider.instance.get(
            instrumentationName: PulseOtelConstants.INSTRUMENTATION_SCOPE,
            instrumentationVersion: PulseOtelConstants.INSTRUMENTATION_VERSION
        )
    }
    
    private static let spanStore = NSMutableDictionary()
    private static let spanStoreQueue = DispatchQueue(label: "com.pulse.spanstore")
    
    @objc(startSpan:inheritContext:attributes:)
    public static func startSpan(name: String, inheritContext: Bool, attributes: NSDictionary?) -> String {
        let builder = getTracer().spanBuilder(spanName: name)
            .setSpanKind(spanKind: SpanKind.internal)

        if let attributes = attributes {
            let swiftAttributes = AttributeValueConverter.convertToSwift(attributes)
            if !swiftAttributes.isEmpty {
                swiftAttributes.forEach { key, value in
                    builder.setAttribute(key: key, value: value)
                }
            }
        }

        // When active, the span will become parent of future spans
        if inheritContext {
            builder.setActive(true)
        }

        let span = builder.startSpan()
        
        let spanId = UUID().uuidString
        spanStoreQueue.sync {
            spanStore[spanId] = span
        }
        
        return spanId
    }
    
    @objc(endSpan:statusCode:)
    public static func endSpan(spanId: String, statusCode: String?) {
        spanStoreQueue.sync {
            guard let span = spanStore[spanId] as? Span else { return }
            
            if let statusCode = statusCode {
                let status = statusCode.uppercased()
                if status == "OK" {
                    span.status = .ok
                } else if status == "ERROR" {
                    span.status = .error(description: "")
                } else {
                    span.status = .unset
                }
            } else {
                span.status = .unset
            }
            
            span.end()
            spanStore.removeObject(forKey: spanId)
        }
    }
    
    @objc(addEvent:name:attributes:)
    public static func addEvent(spanId: String, name: String, attributes: NSDictionary?) {
        spanStoreQueue.sync {
            guard let span = spanStore[spanId] as? Span else { return }
            
            if let attributes = attributes {
                let swiftAttributes = AttributeValueConverter.convertToSwift(attributes)
                if !swiftAttributes.isEmpty {
                    span.setAttributes(swiftAttributes)
                }
            }
            
            span.addEvent(name: name)
        }
    }
    
    @objc(setAttributes:attributes:)
    public static func setAttributes(spanId: String, attributes: NSDictionary?) {
        spanStoreQueue.sync {
            guard let span = spanStore[spanId] as? Span else { return }
            
            if let attributes = attributes {
                let swiftAttributes = AttributeValueConverter.convertToSwift(attributes)
                if !swiftAttributes.isEmpty {
                    span.setAttributes(swiftAttributes)
                }
            }
        }
    }
    
    @objc(recordException:errorMessage:stackTrace:)
    public static func recordException(spanId: String, errorMessage: String, stackTrace: String?) {
        spanStoreQueue.sync {
            guard let span = spanStore[spanId] as? Span else { return }
            
            let error = NSError(domain: "PulseSDK", code: -1, userInfo: [
                NSLocalizedDescriptionKey: errorMessage
            ])
            
            span.recordException(error, attributes: [:])
            span.setAttribute(key: PulseOtelConstants.ATTR_ERROR_MESSAGE, value: OpenTelemetryApi.AttributeValue.string(errorMessage))
            
            if let stackTrace = stackTrace, !stackTrace.isEmpty {
                span.setAttribute(key: PulseOtelConstants.ATTR_ERROR_STACK, value: OpenTelemetryApi.AttributeValue.string(stackTrace))
            }
        }
    }
    
    @objc(discardSpan:)
    public static func discardSpan(spanId: String) {
        spanStoreQueue.sync {
            guard let span = spanStore[spanId] as? Span else { return }
            span.setAttribute(key: "pulse.internal", value: OpenTelemetryApi.AttributeValue.bool(true))
            span.end()
            spanStore.removeObject(forKey: spanId)
        }
    }
    
}

File: pulse-react-native-otel/android/src/main/java/com/pulsereactnativeotel/PulseReactNativeOtelTracer.kt
package com.pulsereactnativeotel

import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableType
import com.pulse.android.sdk.PulseSDK
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.trace.Span
import io.opentelemetry.api.trace.SpanKind
import io.opentelemetry.api.trace.StatusCode
import io.opentelemetry.api.trace.Tracer
import io.opentelemetry.context.Scope
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

object PulseReactNativeOtelTracer {

    private val tracer: Tracer by lazy {
        PulseSDK.INSTANCE.getOtelOrThrow()
            .getOpenTelemetry()
            .tracerProvider
            .tracerBuilder(PulseOtelConstants.INSTRUMENTATION_SCOPE)
            .build()
    }

    private val idToSpan = ConcurrentHashMap<String, Span>()
    private val idToScope = ConcurrentHashMap<String, Scope>()

    fun startSpan(name: String, inheritContext: Boolean, attributes: ReadableMap?): String {
        val span = tracer.spanBuilder(name)
            .setSpanKind(SpanKind.INTERNAL)
            .startSpan()

        attributes?.applyTo(span)

        val id = UUID.randomUUID().toString()
        idToSpan[id] = span
        
        if (inheritContext) {
            val scope = span.makeCurrent()
            idToScope[id] = scope
        }

        return id
    }

    fun addEvent(spanId: String, name: String, attributes: ReadableMap?) {
        idToSpan[spanId]?.let { span ->
            attributes?.applyTo(span)
            span.addEvent(name)
        }
    }

    fun setAttributes(spanId: String, attributes: ReadableMap?) {
        idToSpan[spanId]?.let { span ->
            attributes?.applyTo(span)
        }
    }

    fun recordException(spanId: String, errorMessage: String, stackTrace: String?) {
        idToSpan[spanId]?.let { span ->
            val throwable = RuntimeException(errorMessage)
            span.recordException(throwable)
            span.setAttribute(
                AttributeKey.stringKey(PulseOtelConstants.ATTR_ERROR_MESSAGE),
                errorMessage
            )

            stackTrace?.takeIf { it.isNotEmpty() }?.let {
                span.setAttribute(
                    AttributeKey.stringKey(PulseOtelConstants.ATTR_ERROR_STACK),
                    it
                )
            }
        }
    }

    fun endSpan(spanId: String, statusCode: String?) {
        idToSpan.remove(spanId)?.let { span ->
            when (statusCode?.uppercase()) {
                "OK" -> span.setStatus(StatusCode.OK)
                "ERROR" -> span.setStatus(StatusCode.ERROR)
                else -> span.setStatus(StatusCode.UNSET)
            }
            span.end()
        }
        idToScope.remove(spanId)?.close()
    }

    fun discardSpan(spanId: String) {
        idToSpan.remove(spanId)?.let { span ->
            span.setAttribute(
                AttributeKey.booleanKey("pulse.internal"),
                true
            )
            span.end()
        }
        idToScope.remove(spanId)?.close()
    }

    private fun ReadableMap.applyTo(span: Span) {
        entryIterator.forEach { (key, value) ->
            when (value) {
                is String -> span.setAttribute(AttributeKey.stringKey(key), value)
                is Boolean -> span.setAttribute(AttributeKey.booleanKey(key), value)
                is Number -> span.setAttribute(AttributeKey.doubleKey(key), value.toDouble())
                is ReadableArray -> {
                    applyArrayAttribute(span, key, value)
                }
                else -> {
                    span.setAttribute(AttributeKey.stringKey(key), value.toString())
                }
            }
        }
    }

    private fun applyArrayAttribute(span: Span, key: String, array: ReadableArray) {
        if (array.size() == 0) {
            span.setAttribute(AttributeKey.stringArrayKey(key), emptyList())
            return
        }

        when (array.getType(0)) {
            ReadableType.String -> {
                span.setAttribute(AttributeKey.stringArrayKey(key), array.toArrayList() as List<String>)
            }
            ReadableType.Number -> {
                span.setAttribute(AttributeKey.doubleArrayKey(key), array.toArrayList() as List<Double>)
            }
            ReadableType.Boolean -> {
                span.setAttribute(AttributeKey.booleanArrayKey(key), array.toArrayList() as List<Boolean>)
            }
            else -> {
                span.setAttribute(AttributeKey.stringKey(key), array.toString())
            }
        }
    }
}


